### PR TITLE
feat: combineCols()

### PR DESCRIPTION
**Pull request by keviny2**
pushed original PR (https://github.com/BiocPy/SummarizedExperiment/pull/9) to a branch on main repo and merged changes from master

### CODE REVIEW

**Comment by jkanche** [comment]
LOL this has 60 commits?

**Comment by jkanche** [comment]
nvm, this is going to be squashed and merged 👍🏽

**Comment by keviny2** [comment]
> LOL this has 60 commits?

~20 for the merge with master and ~40 for me 😳

**Comment by jkanche** [comment]
- [ ] missing tutorial in the docs

I'll keep editing this comment as i go

File: src/summarizedexperiment/BaseSE.py [hunk 1]
```diff
@@ -536,3 +543,62 @@ def toAnnData(
...
+                - if any of the provided objects are not "SummarizedExperiment".
+            ValueError:
+                - if there are null or duplicated row names (useNames=True)
+                - if all objects do not have the same number of rows (useNames=False)
+
+        Returns:
+            BaseSE: combined `SummarizedExperiment` object.
+        """
+
+        if len(experiments) == 1 and isinstance(experiments[0], list):
```
Comment (jkanche) [inline src/summarizedexperiment/BaseSE.py:575]: trying to understand this, why would `experiments[0]` be a list?

are we trying to fix for `se.combineCols([se2, se3])`?

Comment (keviny2) [reply src/summarizedexperiment/BaseSE.py:575]: yep, that's exactly why

Comment (jkanche) [reply src/summarizedexperiment/BaseSE.py:575]: we should not be compensating for users mistakes :) 

i'm gonna modify so that it throws an error

Comment (keviny2) [reply src/summarizedexperiment/BaseSE.py:575]: do we not want to allow the user to also pass in a list of SEs?

File: src/summarizedexperiment/utils/_validators.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        any_duplicates = len(df.index) != len(set(df.index))
+        return (not any_null) and (not any_duplicates)
+
+    is_valid_names = all(
+        [_validate_single_df(getattr(se, experiment_attribute)) for se in ses]
+    )
+    if not is_valid_names:
+        raise ValueError(
+            "at least one input `SummarizedExperiment` has null or duplicated row names"
+        )
```
Comment (jkanche) [inline src/summarizedexperiment/utils/_validators.py:42]: not really, this function is again assuming all rowdata are pandas data frames

File: src/summarizedexperiment/dispatchers/combiners.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    left: pd.DataFrame, right: pd.DataFrame, ignore_names: bool, prefer_left: bool
+) -> pd.DataFrame:
+    """Combine dataframes with the appropriate method.
+
+    Args:
+        left (pd.DataFrame): a dataframe.
+        right (pd.DataFrame): a dataframe.
+        ignore_names (bool): if `True`, index names will be ignored when combining.
+        prefer_left (bool): if `True`, only null values in the left dataframe
+            will be populated by values from the right dataframe.
```
Comment (jkanche) [inline src/summarizedexperiment/dispatchers/combiners.py:32]: this should really be called `ignore_na` or something like that

Comment (jkanche) [reply src/summarizedexperiment/dispatchers/combiners.py:32]: I feel this function can be way more simpler, although i might be missing something here. 

```python
def combine_dataframes(
    *frames: pd.DataFrame,
    ignore_names: bool = False,
    axis: int = 0,
) -> pd.DataFrame:
    return pd.concat(frames, ignore_index=ignore_names, axis=axis)
```

e.g:

```python
df1 = pd.DataFrame(
    {
        "A": ["A0", "A1", "A2", "A3"],
        "B": ["B0", "B1", "B2", "B3"],
        "C": ["C0", "C1", "C2", "C3"],
        "D": ["D0", "D1", "D2", "D3"],
    },
    index=[0, 1, 2, 3],
)

df3 = pd.DataFrame(
    {
        "A": ["A8", "A9", "A10", "A11"],
        "B": ["B8", "B9", "B10", "B11"],
        "G": ["G8", "G9", "G10", "G11"],
        "H": ["H8", "H9", "H10", "H11"],
    },
    index=[0, 1, 6, 7],
)


frames = [df1, df3]

row = pd.concat(frames, ignore_index=True, axis=0)
row_ignore_index = pd.concat(frames, ignore_index=False, axis=0)

col = pd.concat(frames, ignore_index=True, axis=1)
col_ignore_index = pd.concat(frames, ignore_index=True, axis=1)
```

Comment (jkanche) [reply src/summarizedexperiment/dispatchers/combiners.py:32]: may be i need an example of why we need to do `combine_first` and have the `prefer_left` param.

**Comment by jkanche** [comment]
we should try to reuse the same test usecases - https://github.com/Bioconductor/SummarizedExperiment/blob/devel/inst/unitTests/test_combine-methods.R

File: src/summarizedexperiment/dispatchers/combiners.py (hunk 3 above)
Comment (keviny2) [reply src/summarizedexperiment/dispatchers/combiners.py:32]: if `useNames=False`, `combineCols` does not perform any concatenation on the rows. so if we call `df1.combineCols(df3, useNames=False)`, the new row names would simply be the row names of the 'first' dataframe - in our case, `[0, 1, 2, 3]`. the example you gave is kind of confusing (not sure why `ignore_index` is different for `row_ignore_index` and `col_ignore_index`), but regardless, pd.concat will always increase the number of rows, which isn't what we want when `useNames=False`. here's some R code using Bioconductor's SummarizedExperiment:

```
library(SummarizedExperiment)

se <- SummarizedExperiment(list(counts=matrix(rpois(50, 10), ncol=10)))
colData(se)$A <- 1L
rowData(se)$A <- 1
rownames(se) <- letters[1:5]

se2 <- SummarizedExperiment(list(counts=matrix(rpois(50, 10), ncol=10), 
                                 normalized=matrix(rnorm(50), ncol=10)))
colData(se2)$A <- 2L
colData(se2)$B <- 3
rowData(se2)$B <- "B"
rownames(se2) <- letters[3:7]

stuff <- combineCols(se, se2, use.names = FALSE)
```

Comment (jkanche) [reply src/summarizedexperiment/dispatchers/combiners.py:32]: I believe this is already fixed in the #17. I added a couple more to keep the index of the first object.

**Pull request merged**